### PR TITLE
Extension for Issue #170

### PR DESCRIPTION
**Pull request by galovics**
Extension to support the test class exclusions based on the test class RunWith annotation.

Now the configuration is extended with the excludedRunners property which is a collection of runners.

### CODE REVIEW

File: pitest/src/main/java/org/pitest/classpath/CodeSource.java [hunk 1]
```diff
@@ -57,7 +63,9 @@ public CodeSource(final ProjectClassPaths classPath,
   public List<ClassInfo> getTests() {
     return flatMap(this.classPath.test(), nameToClassInfo()).filter(
         and(isWithinATestClass(), isIncludedClass(),
-            not(ClassInfo.matchIfAbstract())));
+            not(ClassInfo.matchIfAbstract()),
+            not(ClassInfo.matchIfAnnotationPresentWithValueString(RunWith.class,
```
Comment (KyleRogers) [inline pitest/src/main/java/org/pitest/classpath/CodeSource.java]: I don't think it's a good idea to check this in CodeSource. I think we should move that to JUnitTestClassIdentifier.isIncludedClass()

This is JUnit specific behaviour and shouldn't be executed in TestNG setups. I'm surprised that it seems to run inside a TestNG setup. But Annotations are usually treated differently by the class loader.

**Comment by KyleRogers** [comment]
I'm missing an integration test or acceptance test that demonstrates that the new feature is working as expected. I'm sorry I had not the time to find a suitable Test class to point you to.

File: pitest/src/main/java/org/pitest/classpath/CodeSource.java (hunk 1 above)
Comment (galovics) [inline pitest/src/main/java/org/pitest/classpath/CodeSource.java]: I'll check it out, thanks!

**Comment by galovics** [comment]
Because I haven't added any tests. First, I wanted someone to review the code that this is the direction or not. After I fixed the mentioned issue, I'll add some. :-)

**Comment by galovics** [comment]
The logic went to the JUnitTestClassIdentifier as mentioned plus added tests to verify the functionality.

**Comment by galovics** [comment]
any update here? :-)

File: pitest/src/main/java/org/pitest/junit/JUnitTestClassIdentifier.java [hunk 2]
```diff
@@ -39,7 +42,16 @@ public boolean isATestClass(final ClassInfo a) {
 
   @Override
   public boolean isIncluded(ClassInfo a) {
-    return isIncludedCategory(a) && !isExcludedCategory(a);
+    return isIncludedCategory(a) && !isExcludedCategory(a) && isNotRanWithExcludedRunner(a);
+  }
+
+  private boolean isNotRanWithExcludedRunner(ClassInfo a) {
+    String runWith = getRunWithAnnotationValue(a);
+    return this.excludedRunners.isEmpty() || !this.excludedRunners.contains(runWith);
```
Comment (hcoles) [inline pitest/src/main/java/org/pitest/junit/JUnitTestClassIdentifier.java]: Why is the additional call to isEmpty required here?

Comment (galovics) [inline pitest/src/main/java/org/pitest/junit/JUnitTestClassIdentifier.java]: To be honest, I don't really know why I put that there. Probably it was part of my previous implementation. :-)

Gonna remove it.

**Comment by hcoles** [comment]
I had a quick scan through the diffs last night and it seems reasonable enough, but what is the use case this is supporting? What's the rationale behind filtering out tests by runner?

**Comment by galovics** [comment]
Please check [this](https://github.com/hcoles/pitest/issues/170#issuecomment-230692754) thread from this comment

**Comment by hcoles** [comment]
I don't think we need to worry too much about changing the api - we'd up the version to 1.2.0 but in reality I don't think anyone will really care.

My main concern about this PR is I'm not convinced the functionality will be useful for most people. My preference would be implement this via a plugin rather than as a feature of the main codebase, however I don't think a suitable extension point currently exists.

**Comment by galovics** [comment]
This can be implemented as a plugin of course. However, if there is no such extension point yet, we can include this in the main codebase and when the plugin support is there, we can migrate it.

**Comment by galovics** [comment]
Thank you for merging it. May I ask you when do you plan the next release? :-)

**Pull request merged**